File: 2025/day06/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Math::BigInt;

# Read input file
my $input_file = "../input.txt";
open(my $fh, '<', $input_file) or die "Cannot open $input_file: $!";
my @lines = <$fh>;
close($fh);

# Remove trailing newlines but keep the content
chomp(@lines);

sub parse_problems {
    my ($lines_ref) = @_;
    my @lines = @$lines_ref;

    return () if @lines == 0;

    # Find the operator row (last non-empty row with only +, *, and spaces)
    my $op_row_idx = $#lines;
    while ($op_row_idx >= 0) {
        my $line = $lines[$op_row_idx];
        last if $line =~ /\S/ && $line =~ /^[+*\s]+$/;
        $op_row_idx--;
    }

    return () if $op_row_idx < 0;

    my $op_row = $lines[$op_row_idx];
    my @number_rows = @lines[0..$op_row_idx-1];

    # Find max width
    my $max_width = 0;
    for my $line (@lines) {
        my $len = length($line);
        $max_width = $len if $len > $max_width;
    }

    # Pad all rows to the same width
    my @padded_number_rows = map { sprintf("%-${max_width}s", $_) } @number_rows;
    my $padded_op_row = sprintf("%-${max_width}s", $op_row);

    # Find problem boundaries by looking for columns that are all spaces
    my @problems;
    my $col = 0;

    while ($col < $max_width) {
        # Skip separator columns (all spaces)
        while ($col < $max_width) {
            my $is_all_space = 1;
            for my $row (@padded_number_rows) {
                if (substr($row, $col, 1) ne ' ') {
                    $is_all_space = 0;
                    last;
                }
            }
            last unless $is_all_space && substr($padded_op_row, $col, 1) eq ' ';
            $col++;
        }

        last if $col >= $max_width;

        # Find the end of this problem
        my $start_col = $col;
        while ($col < $max_width) {
            # Check if this is a separator column
            my $is_separator = 1;
            for my $row (@padded_number_rows) {
                if (substr($row, $col, 1) ne ' ') {
                    $is_separator = 0;
                    last;
                }
            }
            $is_separator = 0 if substr($padded_op_row, $col, 1) ne ' ';
            last if $is_separator;
            $col++;
        }

        my $end_col = $col;

        # Extract numbers and operator for this problem
        my @numbers;
        for my $row (@padded_number_rows) {
            my $num_str = substr($row, $start_col, $end_col - $start_col);
            $num_str =~ s/^\s+|\s+$//g;  # trim
            if ($num_str ne '') {
                push @numbers, Math::BigInt->new($num_str);
            }
        }

        my $op_str = substr($padded_op_row, $start_col, $end_col - $start_col);
        $op_str =~ s/^\s+|\s+$//g;  # trim
        if ($op_str ne '' && @numbers) {
            push @problems, {
                numbers => \@numbers,
                op => $op_str
            };
        }
    }

    return @problems;
}

sub solve_problem {
    my ($numbers_ref, $op) = @_;
    my @numbers = @$numbers_ref;

    if ($op eq '+') {
        my $result = Math::BigInt->new(0);
        for my $n (@numbers) {
            $result->badd($n);
        }
        return $result;
    } elsif ($op eq '*') {
        my $result = Math::BigInt->new(1);
        for my $n (@numbers) {
            $result->bmul($n);
        }
        return $result;
    }
    return Math::BigInt->new(0);
}

sub part1 {
    my @problems = parse_problems(\@lines);
    my $total = Math::BigInt->new(0);

    for my $problem (@problems) {
        my $result = solve_problem($problem->{numbers}, $problem->{op});
        $total->badd($result);
    }

    return $total;
}

sub parse_problems_part2 {
    my ($lines_ref) = @_;
    my @lines = @$lines_ref;

    return () if @lines == 0;

    # Find the operator row (last non-empty row with only +, *, and spaces)
    my $op_row_idx = $#lines;
    while ($op_row_idx >= 0) {
        my $line = $lines[$op_row_idx];
        last if $line =~ /\S/ && $line =~ /^[+*\s]+$/;
        $op_row_idx--;
    }

    return () if $op_row_idx < 0;

    my $op_row = $lines[$op_row_idx];
    my @number_rows = @lines[0..$op_row_idx-1];

    # Find max width
    my $max_width = 0;
    for my $line (@lines) {
        my $len = length($line);
        $max_width = $len if $len > $max_width;
    }

    # Pad all rows to the same width
    my @padded_number_rows = map { sprintf("%-${max_width}s", $_) } @number_rows;
    my $padded_op_row = sprintf("%-${max_width}s", $op_row);

    # Find problem boundaries by looking for columns that are all spaces
    my @problems;
    my $col = 0;

    while ($col < $max_width) {
        # Skip separator columns (all spaces)
        while ($col < $max_width) {
            my $is_all_space = 1;
            for my $row (@padded_number_rows) {
                if (substr($row, $col, 1) ne ' ') {
                    $is_all_space = 0;
                    last;
                }
            }
            last unless $is_all_space && substr($padded_op_row, $col, 1) eq ' ';
            $col++;
        }

        last if $col >= $max_width;

        # Find the end of this problem
        my $start_col = $col;
        while ($col < $max_width) {
            # Check if this is a separator column
            my $is_separator = 1;
            for my $row (@padded_number_rows) {
                if (substr($row, $col, 1) ne ' ') {
                    $is_separator = 0;
                    last;
                }
            }
            $is_separator = 0 if substr($padded_op_row, $col, 1) ne ' ';
            last if $is_separator;
            $col++;
        }

        my $end_col = $col;

        # For Part 2: Read columns right-to-left, each column forms a number
        # reading top-to-bottom as most-to-least significant digit
        my @numbers;
        for (my $c = $end_col - 1; $c >= $start_col; $c--) {  # Right to left
            my @digits;
            for my $row (@padded_number_rows) {
                my $ch = substr($row, $c, 1);
                if ($ch =~ /\d/) {
                    push @digits, $ch;
                }
            }
            if (@digits) {
                # Join digits to form number (top=most significant, bottom=least)
                my $num_str = join('', @digits);
                push @numbers, Math::BigInt->new($num_str);
            }
        }

        my $op_str = substr($padded_op_row, $start_col, $end_col - $start_col);
        $op_str =~ s/^\s+|\s+$//g;  # trim
        if ($op_str ne '' && @numbers) {
            push @problems, {
                numbers => \@numbers,
                op => $op_str
            };
        }
    }

    return @problems;
}

sub part2 {
    my @problems = parse_problems_part2(\@lines);
    my $total = Math::BigInt->new(0);

    for my $problem (@problems) {
        my $result = solve_problem($problem->{numbers}, $problem->{op});
        $total->badd($result);
    }

    return $total;
}

# Main execution
print "Part 1: ", part1(), "\n";
print "Part 2: ", part2(), "\n";

File: 2025/day06/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

// Parse input
const lines = input.split('\n');

function parseProblems(lines) {
  if (lines.length === 0) return [];

  // Find the operator row (last non-empty row with only +, *, and spaces)
  let opRowIdx = lines.length - 1;
  while (opRowIdx >= 0 && (!lines[opRowIdx].trim() ||
         ![...lines[opRowIdx]].every(c => c === '+' || c === '*' || c === ' '))) {
    opRowIdx--;
  }

  if (opRowIdx < 0) return [];

  const opRow = lines[opRowIdx];
  const numberRows = lines.slice(0, opRowIdx);

  // Find max width
  const maxWidth = Math.max(...lines.map(l => l.length));

  // Pad all rows to the same width
  const paddedNumberRows = numberRows.map(row => row.padEnd(maxWidth));
  const paddedOpRow = opRow.padEnd(maxWidth);

  // Find problem boundaries by looking for columns that are all spaces
  const problems = [];
  let col = 0;

  while (col < maxWidth) {
    // Skip separator columns (all spaces)
    while (col < maxWidth && paddedNumberRows.every(row => row[col] === ' ') && paddedOpRow[col] === ' ') {
      col++;
    }

    if (col >= maxWidth) break;

    // Find the end of this problem
    const startCol = col;
    while (col < maxWidth) {
      // Check if this is a separator column
      const isSeparator = paddedNumberRows.every(row => row[col] === ' ') && paddedOpRow[col] === ' ';
      if (isSeparator) break;
      col++;
    }

    const endCol = col;

    // Extract numbers and operator for this problem
    const numbers = [];
    for (const row of paddedNumberRows) {
      const numStr = row.slice(startCol, endCol).trim();
      if (numStr) {
        numbers.push(BigInt(numStr));
      }
    }

    const opStr = paddedOpRow.slice(startCol, endCol).trim();
    if (opStr && numbers.length > 0) {
      problems.push({ numbers, op: opStr });
    }
  }

  return problems;
}

function solveProblem(numbers, op) {
  if (op === '+') {
    return numbers.reduce((a, b) => a + b, 0n);
  } else if (op === '*') {
    return numbers.reduce((a, b) => a * b, 1n);
  }
  return 0n;
}

// Part 1
function part1() {
  const problems = parseProblems(lines);
  let total = 0n;
  for (const { numbers, op } of problems) {
    total += solveProblem(numbers, op);
  }
  return total.toString();
}

function parseProblemsPart2(lines) {
  if (lines.length === 0) return [];

  // Find the operator row (last non-empty row with only +, *, and spaces)
  let opRowIdx = lines.length - 1;
  while (opRowIdx >= 0 && (!lines[opRowIdx].trim() ||
         ![...lines[opRowIdx]].every(c => c === '+' || c === '*' || c === ' '))) {
    opRowIdx--;
  }

  if (opRowIdx < 0) return [];

  const opRow = lines[opRowIdx];
  const numberRows = lines.slice(0, opRowIdx);

  // Find max width
  const maxWidth = Math.max(...lines.map(l => l.length));

  // Pad all rows to the same width
  const paddedNumberRows = numberRows.map(row => row.padEnd(maxWidth));
  const paddedOpRow = opRow.padEnd(maxWidth);

  // Find problem boundaries by looking for columns that are all spaces
  const problems = [];
  let col = 0;

  while (col < maxWidth) {
    // Skip separator columns (all spaces)
    while (col < maxWidth && paddedNumberRows.every(row => row[col] === ' ') && paddedOpRow[col] === ' ') {
      col++;
    }

    if (col >= maxWidth) break;

    // Find the end of this problem
    const startCol = col;
    while (col < maxWidth) {
      // Check if this is a separator column
      const isSeparator = paddedNumberRows.every(row => row[col] === ' ') && paddedOpRow[col] === ' ';
      if (isSeparator) break;
      col++;
    }

    const endCol = col;

    // For Part 2: Read columns right-to-left, each column forms a number
    // reading top-to-bottom as most-to-least significant digit
    const numbers = [];
    for (let c = endCol - 1; c >= startCol; c--) {
      const digits = [];
      for (const row of paddedNumberRows) {
        const ch = row[c];
        if (ch >= '0' && ch <= '9') {
          digits.push(ch);
        }
      }
      if (digits.length > 0) {
        const num = BigInt(digits.join(''));
        numbers.push(num);
      }
    }

    const opStr = paddedOpRow.slice(startCol, endCol).trim();
    if (opStr && numbers.length > 0) {
      problems.push({ numbers, op: opStr });
    }
  }

  return problems;
}

// Part 2
function part2() {
  const problems = parseProblemsPart2(lines);
  let total = 0n;
  for (const { numbers, op } of problems) {
    total += solveProblem(numbers, op);
  }
  return total.toString();
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());
